core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -1448,7 +1448,7 @@ private[spark] object MapOutputTracker extends Logging {
     // TODO: improve push based shuffle to read partial merged blocks satisfying the start/end
     // TODO: map indexes
     if (mergeStatuses.exists(_.nonEmpty) && startMapIndex == 0
-      && endMapIndex == mapStatuses.length) {
+      && endMapIndex == mapStatuses.length && endPartition - startPartition == 1) {
```

Comment:
I think this enhanced condition is enough to fix the issue. Why do we still have to disable the batch fetch?

Reply:
I agree - this condition should be sufficient.
With both conditions, we will end up not using batch fetch or push based shuffle for coalasced partitions, right ?

Reply:
Not push-based shuffle, but batch fetch is still doable, right?

Reply:
Why not keep the behavior of push-based shuffle and disable the behavior of batch fetch when partition coalesce happens?

Reply:
> My concern with only keeping this condition (endPartition - startPartition == 1) instead of disable batch fetch is that it will effectively disable push-based shuffle in 3.2.0 even if the push.enabled flag is set to true.

IIUC, push-based shuffle currently can only work with one reduce partition, no? And I didn't see `endPartition - startPartition == 1` disabled push-based shuffle for any cases except "partition coalesce".

Reply:
Push based shuffle can work with multiple partitions per reducer. 
What it cannot work with is when one shuffle partition gets split between multiple reducers in the case of skew handling.

Reply:
> Push based shuffle can work with multiple partitions per reducer.

Yeah, I get it after second thinking. But given that, currently, there's actually no case that has multiple partitions, would  
`endPartition - startPartition == 1` be a simpler fix to go?

> Why not keep the behavior of push-based shuffle and disable the behavior of batch fetch when partition coalesce happens?

For this way, I'm afraid we have to change the interface `ShuffleManager.getReader`, which could involve more changes.

Reply:
> For this way, I'm afraid we have to change the interface ShuffleManager.getReader, which could involve more changes.
I don't think changing the interface to `ShuffleManager.getReader` is needed.
As demonstrated in this PR inside `BlockStoreShuffleReader`, we can disable batch fetch when push-based shuffle is enabled and it's an equally simple fix.

The default configs in 3.2.0 will initiate partition coalesce if the size per shuffle partition is less than 1MB, and people might want to enable partition coalesce to fix the many small files problems.
These valid reasons where partition coalesce is preferred to be enabled might make it even trickier to reason with when push-based shuffle actually gets enabled.

Block batching and push-based shuffle could potentially work together, i.e. we could batch blocks together to be pushed to create merged shuffle files that span across multiple shuffle partitions and enable push-based shuffle and batch fetching to properly work together.
However, for the sake of unblocking 3.2.0 release, we need a much quicker resolution to this issue.
For that, when the flags for AQE, partition coalesce, batch fetch, and push-based shuffle are all set to true, we will have to either disable push-based shuffle and leave batch fetch as is with partition coalesce, or we can disable batch fetch and leave push-based shuffle as is with partition coalesce.
I prefer the 2nd option, because it makes it more consistent for enabling the flag of push-based shuffle and push-based shuffle is a more thorough solution for the problem that batch fetch is trying to solve.

Reply:
> As demonstrated in this PR inside BlockStoreShuffleReader, we can disable batch fetch when push-based shuffle is enabled and it's an equally simple fix.

Ok, so you mean to disable batch fetch for all the cases when push-based shuffle rather than the partition coalesce case only. Then, of course, the change is also simple.

Reply:
I believe this should suffice:
```
    if (mergeStatuses.exists(_.nonEmpty) && mergeStatuses.exists(_.exists(_ != null))
      && startMapIndex == 0
      && endMapIndex == mapStatuses.length) {
```

We can remove the changes to `BlockStoreShuffleReader`.
Thoughts ?

Reply:
To elaborate:

* If push based shuffle is enabled for a stage we want to disable block fetch.
  * The check in `BlockStoreShuffleReader` does this.
* If block fetch is enabled, we want to make sure it always goes through the `else` condition here.
  * block fetch assumes ordering of id's which get violated in the `if` block here.

For the case of `Utils.isPushBasedShuffleEnabled == true` and `BlockStoreShuffleReader.shouldBatchFetch == true`, we have following cases:

* `shuffleStage.shuffleMergeEnabled == false`
  * `mergeStatuses != null` and `mergeStatuses.length == numReducers` though `mergeStatuses` has only nulls.
  * With current PR, `doBatchFetch` == true, but we take the `if` condition in `convertMapStatuses` - which results in incorrect ordering of blocks.
  * With the proposed change above, we will take else block.
    * This is the root cause of the hang.
  * Additional change - we always take `else` path even for push based shuffle in this case.
    * This should be functionally equivalent - since we can now rely on block fetch when push based shuffle does not kick in (stage retry for example) - so better behavior; but more importantly, should not cause issues.
* `shuffleStage.shuffleMergeEnabled == true`
  * With this PR, block fetch disabled.
  * With push based shuffle having merged blocks, it takes the `if` block - else it takes `else` block (for direct fetch of unmerged blocks).
    * No behavior change : other than disabling block fetch.
    * This prevents duplicate data fetch issue.

Reply:
@Ngone51 
Just take a look at the usage of `PartialMapperPartitionSpec` and `CoalescedMapperPartitionSpec` for the local shuffle read optimization.
When the local shuffle read optimization kicks in, it indeed makes sense to not do push-based shuffle for that special case since it's supposed to be a local shuffle read for changing partitions of the table after AQE-converted bhj.
In these cases, because the reducer for the local shuffle is reading from a subset of the mappers, push-based shuffle cannot happen in the first place.
However, the current condition for disabling batch fetch in `BlockStoreShuffleReader` would only check if push-based shuffle is enabled for the given shuffle dependency, and it will still disable batch fetch even if push-based shuffle cannot happen because the reducer is not fetching blocks from all the mappers.
We are evaluating passing this additional information to `BlockStoreShuffleReader` as well so that we can properly disable batch fetch only for cases of partition coalesce but not for local shuffle read.
The PR will be updated with this change once validated.

Reply:
We don’t need to touch ShuffleManager API. 
In fact, we don’t need to touch any API at all after some more internal discussions. 
@zhouyejoe will update the PR after some more tests to show the change.

Reply:
Updated the PR, please help review. Thanks.

Reply:
UT to be added